exporter/awsxrayexporter/translator/segment.go

```diff
@@ -0,0 +1,323 @@
+	tracepb "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1"
+	"github.com/golang/protobuf/ptypes/timestamp"
+	semconventions "github.com/open-telemetry/opentelemetry-collector/translator/conventions"
+	tracetranslator "github.com/open-telemetry/opentelemetry-collector/translator/trace"
+)
+
+// AWS X-Ray acceptable values for origin field.
+const (
+	OriginEC2 = "AWS::EC2::Instance"
+	OriginECS = "AWS::ECS::Container"
+	OriginEB  = "AWS::ElasticBeanstalk::Environment"
+)
+
+var (
+	zeroSpanID = []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}
+	r          = rand.New(rand.NewSource(time.Now().UnixNano())) // random, not secure
+	mutex      = &sync.Mutex{}
+)
+
+var (
+	// reInvalidSpanCharacters defines the invalid letters in a span name as per
+	// https://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html
+	reInvalidSpanCharacters = regexp.MustCompile(`[^ 0-9\p{L}N_.:/%&#=+,\-@]`)
+	// reInvalidAnnotationCharacters defines the invalid letters in an annotation key as per
+	// https://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html
+	reInvalidAnnotationCharacters = regexp.MustCompile(`[^a-zA-Z0-9_]`)
+)
+
+const (
+	// defaultSpanName will be used if there are no valid xray characters in the span name
+	defaultSegmentName = "span"
+	// maxSegmentNameLength the maximum length of a Segment name
+	maxSegmentNameLength = 200
+)
+
+const (
+	traceIDLength    = 35 // fixed length of aws trace id
+	identifierOffset = 11 // offset of identifier within traceID
+)
+
+// Segment provides the shape for unmarshalling segment data.
+type Segment struct {
+	// Required
+	TraceID   string  `json:"trace_id,omitempty"`
+	ID        string  `json:"id"`
+	Name      string  `json:"name"`
+	StartTime float64 `json:"start_time"`
+	EndTime   float64 `json:"end_time,omitempty"`
+
+	// Optional
+	InProgress  bool       `json:"in_progress,omitempty"`
+	ParentID    string     `json:"parent_id,omitempty"`
+	Fault       bool       `json:"fault,omitempty"`
+	Error       bool       `json:"error,omitempty"`
+	Throttle    bool       `json:"throttle,omitempty"`
+	Cause       *CauseData `json:"cause,omitempty"`
+	ResourceARN string     `json:"resource_arn,omitempty"`
+	Origin      string     `json:"origin,omitempty"`
+
+	Type         string   `json:"type,omitempty"`
+	Namespace    string   `json:"namespace,omitempty"`
+	User         string   `json:"user,omitempty"`
+	PrecursorIDs []string `json:"precursor_ids,omitempty"`
+
+	HTTP *HTTPData `json:"http,omitempty"`
+	AWS  *AWSData  `json:"aws,omitempty"`
+
+	Service *ServiceData `json:"service,omitempty"`
+
+	// SQL
+	SQL *SQLData `json:"sql,omitempty"`
+
+	// Metadata
+	Annotations map[string]interface{}            `json:"annotations,omitempty"`
+	Metadata    map[string]map[string]interface{} `json:"metadata,omitempty"`
+}
+
+var (
+	writers = newWriterPool(2048)
+)
+
+// MakeSegmentDocumentString converts an Otel Span to an X-Ray Segment and then serialzies to JSON
```

Comment:
This is actually an OpenCensus Span, not Otel Span. Same for function below.

Reply:
Fixed